### PR TITLE
support parallel align

### PR DESCRIPTION
**Pull request by wjwei-handsome**

### CODE REVIEW

File: star-aligner/src/lib.rs [hunk 1]
```diff
@@ -1,14 +1,49 @@
...
+    // collections::HashMap,
+    ffi::{c_char, c_int, CStr, CString},
+    fs::File,
+    io::{BufRead, BufReader},
+    path::{Path, PathBuf},
+    sync::Arc,
+    vec,
+};
+
+pub trait TranscriptomeAligner {
```
Comment (kaizhang) [inline star-aligner/src/lib.rs]: This trait should go to "precellar".

File: star-aligner/src/lib.rs [hunk 2]
```diff
@@ -109,92 +157,211 @@ impl Drop for StarAligner {
...
+    // align single read in parallel
+    pub fn align_reads(
+        &mut self,
+        fqrecs: &mut [fastq::Record],
+        n_threads: usize,
+    ) -> Result<Vec<SamRecord>> {
+        // set thread pool
+        rayon::ThreadPoolBuilder::new()
+            .num_threads(n_threads)
+            .build_global()?;
```
Comment (kaizhang) [inline star-aligner/src/lib.rs]: It is better not to change the global thread pool.  Use `install()` to run it in a local thread pool.

File: star-aligner/src/lib.rs [hunk 3]
```diff
@@ -109,92 +157,211 @@ impl Drop for StarAligner {
...
+
+        // align reads in parallel
+        let records: Vec<SamRecord> = fqrecs_chunks
+            .par_iter()
+            .flat_map(|chunk| {
+                chunk
+                    .iter()
+                    .flat_map(|fqrec| {
+                        // lock the aligner
+                        let mut aligner = aligner.lock().unwrap();
```
Comment (kaizhang) [inline star-aligner/src/lib.rs]: Using `Mutex` means there will be only one single thread holding the aligner and doing the job at a time. You're not using multithreading. You should simply clone the aligner within "flat_map" loop.

File: star-aligner/src/lib.rs [hunk 4]
```diff
@@ -109,92 +157,211 @@ impl Drop for StarAligner {
...
+            .collect();
+
+        Ok(records)
+    }
+
+    pub fn align_read(&mut self, fqrec: &FastqRecord) -> Result<Vec<SamRecord>> {
+        // init a aligner
+        self.aligner = unsafe { star_sys::init_aligner_from_ref(self.index.0.as_ref().index) };
+
+        let mut fq_buf = Vec::new();
```
Comment (kaizhang) [inline star-aligner/src/lib.rs]: It may save some time if you allocate buffers outside this function, and reuse them across all reads from the same chunk.

**Review by kaizhang — CHANGES_REQUESTED** [review]
Please read my comments.

File: star-aligner/src/lib.rs (hunk 1 above)
Comment (wjwei-handsome) [reply star-aligner/src/lib.rs]: OK

**Pull request merged**